src/core/providers/ogr/qgsogrprovidermetadata.cpp

```diff
@@ -1098,6 +1099,93 @@ QList<QgsProviderSublayerDetails> QgsOgrProviderMetadata::querySublayers( const
     }
   }
 
+  const QString path = uriParts.value( QStringLiteral( "path" ) ).toString();
+  const QFileInfo pathInfo( path );
+  if ( flags & Qgis::SublayerQueryFlag::FastScan && ( pathInfo.isFile() || pathInfo.isDir() ) )
+  {
+    // fast scan, so we don't actually try to open the dataset and instead just check the extension alone
+    const QStringList fileExtensions = QgsOgrProviderUtils::fileExtensions();
+    const QStringList dirExtensions = QgsOgrProviderUtils::directoryExtensions();
+
+    const QString suffix = pathInfo.suffix().toLower();
+
+    // allow only normal files or supported directories to continue
+    const bool isOgrSupportedDirectory = pathInfo.isDir() && dirExtensions.contains( suffix );
+    if ( !isOgrSupportedDirectory && !pathInfo.isFile() )
+      return {};
+
+    if ( !fileExtensions.contains( suffix ) && !dirExtensions.contains( suffix ) )
+    {
+      bool matches = false;
+      const QStringList wildcards = QgsOgrProviderUtils::wildcards();
+      for ( const QString &wildcard : wildcards )
+      {
+        const QRegularExpression rx( QRegularExpression::wildcardToRegularExpression( wildcard ), QRegularExpression::CaseInsensitiveOption );
+        if ( rx.match( pathInfo.fileName() ).hasMatch() )
+        {
+          matches = true;
+          break;
+        }
+      }
+      if ( !matches )
+        return {};
+    }
+
+    // these extensions are trivial to read, so there's no need to rely on
```

Comment:
I'm not sure to understand this comment : "trivial to read" ?
And it's odd to see the xlsx extension both in sSkipFastTrackExtensions and sOgrSupportedDbLayersExtensions.

Reply:
The intention here is that any drivers which are cheap to sniff layers from a file should be listed here, so that we know it's not prohibitively expensive to always open them when scanning files (even if they are network drives etc). Do you have any suggestions on additional types which should be added to this list?

Reply:
ok, but the intersection between sSkipFastTrackExtensions and sOgrSupportedDbLayersExtensions should be empty, right ?
And I'm not sure why .shp.zip is there, and not .shp

That's probably nitpicking, but I'm not sure that ods, xls or xlsx are always really cheap to open. If those files were very big (like 1 GB), it could be slow for OGR to figure out the schema as it would have to scan their content completely.

Reply:
> ok, but the intersection between sSkipFastTrackExtensions and sOgrSupportedDbLayersExtensions should be empty, right ?

Not necessarily... In the upcoming commit where I implement FastScan for gdal provider I've moved these too lists to QgsGdalUtils. And in a generic sense I don't see there being an issue with overlap between the lists (when used for other purposes outside of query sublayers, that is!)

> And I'm not sure why .shp.zip is there, and not .shp

I'm not sure either -- I suspect it may have been added as a quick way of getting .shp.zip showing as containers in the browser. But it should be in the other "supported db layers extensions" list, not the inexpensive to scan list. I'll move it.

> That's probably nitpicking, but I'm not sure that ods, xls or xlsx are always really cheap to open. If those files were very big (like 1 GB), it could be slow for OGR to figure out the schema as it would have to scan their content completely.

Hmmm... what would you feel about a `QgsGdalUtils::isCheapToOpen( const QString& extension, long long fileSize )` method? We could return true for all files < some threshold, and have other formats which just require some quick header parsing return true also...

Reply:
> And in a generic sense I don't see there being an issue with overlap between the lists (when used for other purposes outside of query sublayers, that is!)

ok, got it (my point was about the code as it is currently)

> what would you feel about a QgsGdalUtils::isCheapToOpen( const QString& extension, long long fileSize ) method?

That could do it.
Side remark: in an ideal world QGIS should not have to know all this. That should be something available in GDAL.

Reply:
> Side remark: in an ideal world QGIS should not have to know all this. That should be something available in GDAL.

Agreed (as far as i can see, there's no suitable GDAL driver metadata I can use at the moment, is there?). Anyway, a big part of the motivation here is condensing all the multiple parts of QGIS where this logic is scattered through so that we've a better idea what could/should be moved upstream.

Reply:
Ok, I've started on this in https://github.com/qgis/QGIS/pull/44316

Reply:
@rouault 
> Side remark: in an ideal world QGIS should not have to know all this. That should be something available in GDAL.

Was just about to add a DMD metadata flag for drivers which support sublayers, but I see there's already "GDAL_DMD_SUBDATASETS". Does this have the same meaning, or is it a slightly different concept? (I notice drivers like ods/xls/pdf/... aren't returning it, so if it does mean the same thing as "supports layers" then I'll submit a PR adding it for these)

Reply:
> GDAL_DMD_SUBDATASETS

This is for raster drivers that have "child layers", like netCDF or multi-table raster GeoPackage.
Perhasp a GDAL_DCAP_MULTIPLE_VECTOR_LAYERS = "DCAP_MULTIPLE_VECTOR_LAYERS"